### ISSUE TITLE
Strange behaviour with added kwargs to PurePath subclass

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#5889]
**Describe the bug**
Subclassing `PurePath` and adding kwargs to the method as now allowed in python3.12 breaks pyright.

**Code or Screenshots**
```python
from __future__ import annotations

from pathlib import PurePath
from typing import TYPE_CHECKING, Self

if TYPE_CHECKING:
    from _typeshed import StrPath


class ManifestPath(PurePath):
    def __init__(self, *args: StrPath, manifest: int):
        super().__init__()
        self._manifest = manifest

    def with_segments(self, *args: StrPath) -> Self:
        return self.__class__(*args, manifest=self._manifest)
```
<img width="801" alt="image" src="https://github.com/microsoft/pyright/assets/50501825/d04ea176-e89b-4884-b4b6-0c63e8ad3333">
Pylance seems to be reporting the correct signature here but pyright says it is wrong
<img width="477" alt="image" src="https://github.com/microsoft/pyright/assets/50501825/a7787f46-da47-4e59-8baa-6ed2a61d4d01">

**VS Code extension or command-line**
pythonVersion: 3.13
pylance: v2023.8.50

**Comment by erictraut** [comment]
Pyright's behavior is correct here. The class `PurePath` defines a `__new__` method `def __new__(cls, *args: StrPath) -> Self: ...` that doesn't support a keyword parameter named `manifest`. The subclass `ManifestPath` doesn't override this `__new__` method, so the use of such a keyword should generate an error.

It's not clear to me why pylance isn't displaying this error. It correctly displays it when I enable strict-mode type checking (e.g. by adding a `# pyright: strict` comment to the top of the file), but this shouldn't be required. This check is associated with `reportGeneralTypeIssues`, which is enabled with basic type checking.

@debonte, could you help debug this in the context of pylance?

**Comment by Gobot1234** [comment]
In 3.12 though they added support for `**kwargs` in `__new__` so it should be fine.

**Comment by erictraut** [comment]
It looks like this was just recently changed within the `pathlib.pyi` typeshed stub. The typeshed stubs that are bundled with the latest versions of pyright and pylance do not include this.

**Comment by Gobot1234** [comment]
Ah, I was just looking at the typeshed repo, I didn't check pyright's copy. Updating the stubs manually has fixed the issue

**Comment by erictraut** [comment]
That still doesn't explain the difference in behavior between pyright and pylance. I'd like to investigate that further before closing this.

**Comment by erictraut** [comment]
I just pulled the latest typeshed stubs. I normally update them every other pyright release (every two weeks), but I can pull them more often as necessary. The latest changes to `pathlib.pyi` will be included in this week's pyright release.

**Comment by erictraut** [comment]
I just filed a [bug report in the pylance issue tracker](https://github.com/microsoft/pylance-release/issues/4799). I think pyright is doing the right thing here. I can't explain why we're seeing inconsistent behavior with pylance. I'll need someone from the pylance team to investigate the problem.
